### PR TITLE
[mlir][spirv] Support alias/restrict function argument decorations

### LINKED ISSUE #76106: [mlir][spirv] Support function argument decorations
We need to support function argument decorations in `spirv.func` to allow for some pointer types. For example, the spec requires the following:

> If an OpFunctionParameter is a pointer (or array of pointers) in PhysicalStorageBuffer [storage class](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_physical_storage_buffer.html#Storage_Class), then the function parameter must be decorated with exactly one of Aliased or Restrict.

> If an OpFunctionParameter is a pointer (or array of pointers) and its pointee type is a pointer in PhysicalStorageBuffer [storage class](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_physical_storage_buffer.html#Storage_Class), then the function parameter must be decorated with exactly one of AliasedPointer or RestrictPointer.

Note that this is already supported with the `spirv.Variable` op using the following syntax:
```mlir
    %3 = spirv.Variable {aliased_pointer} :
      !spirv.ptr<!spirv.ptr<f32, PhysicalStorageBuffer>, Function>
```

### PR DESCRIPTION
**Pull request by sott0n**
Closes #76106

### CODE REVIEW

File: mlir/lib/Target/SPIRV/Serialization/Serializer.cpp
Comment (kuhar) [inline mlir/lib/Target/SPIRV/Serialization/Serializer.cpp:293]: nit: Write full types instead of `auto` when the type is not obvious based on the RHS only

File: mlir/lib/Target/SPIRV/Serialization/Serializer.cpp
Comment (kuhar) [inline mlir/lib/Target/SPIRV/Serialization/Serializer.cpp:277]:
```suggestion
    if (isa<UnitAttr, DecorationAttr>(attr))
```

File: mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp:994]:
```suggestion
  for (unsigned i = 0, e = funcOp.getNumArguments(); i != e; ++i) {
```

**Review by kuhar — COMMENTED** [review]
Awesome, thanks for working on this @sott0n!
I left some suggestions, mostly on the coding style. I'm not super familiar with the deserializer, so it would be best if @antiagainst could take a look to.

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td
Comment (sott0n) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td:290]: This means that `Aliased` or `Restrict` could be marked, so should i write separate examples like below?

```
    spirv.func @aliased_pointer(%arg0: !spirv.ptr<i32, PhysicalStorageBuffer>,
        { spirv.decoration = #spirv.decoration<Aliased> }) -> () "None" { ... }

    spirv.func @restrict_pointer(%arg0: !spirv.ptr<i32, PhysicalStorageBuffer>,
        { spirv.decoration = #spirv.decoration<Restrict> }) -> () "None" { ... }
```

**Comment by sott0n** [comment]
@kuhar Thanks for your review! I addressed your comments.

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td
Comment (kuhar) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td]: We should have 2 separate test cases like above. But further, we should also have the verifier reject arguments with both decorations.

File: mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp:992]: nit: use the full type, since it's not obvious based on the RHS only

File: mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp:1017]: nit: Use `continue` and remove the `else` here

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (kuhar) [inline mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:381]:
```suggestion
  // e.g. `aliased` -> `spirv.decoration = #spirv.decoration<Aliased>`.
```

File: mlir/test/Target/SPIRV/function-decorations.mlir
Comment (kuhar) [inline mlir/test/Target/SPIRV/function-decorations.mlir:26]:
```suggestion
  // CHECK-LABEL: spirv.func @func_arg_decoration_aliased(%{{.*}}: !spirv.ptr<i32, PhysicalStorageBuffer> {spirv.decoration = #spirv.decoration<Aliased>})
```

File: mlir/test/Target/SPIRV/function-decorations.mlir
Comment (kuhar) [inline mlir/test/Target/SPIRV/function-decorations.mlir:33]:
```suggestion
  // CHECK-LABEL: spirv.func @func_arg_decoration_restrict(%{{.*}}: !spirv.ptr<i32, PhysicalStorageBuffer> {spirv.decoration = #spirv.decoration<Restrict>})
```

File: mlir/test/Target/SPIRV/function-decorations.mlir
Comment (kuhar) [inline mlir/test/Target/SPIRV/function-decorations.mlir:40]:
```suggestion
  // CHECK-LABEL: spirv.func @func_arg_decoration_aliased_pointer(%{{.*}}: !spirv.ptr<!spirv.ptr<i32, PhysicalStorageBuffer>, Generic> {spirv.decoration = #spirv.decoration<AliasedPointer>})
```

File: mlir/test/Target/SPIRV/function-decorations.mlir
Comment (kuhar) [inline mlir/test/Target/SPIRV/function-decorations.mlir:47]:
```suggestion
  // CHECK-LABEL: spirv.func @func_arg_decoration_restrict_pointer(%{{.*}}: !spirv.ptr<!spirv.ptr<i32, PhysicalStorageBuffer>, Generic> {spirv.decoration = #spirv.decoration<RestrictPointer>})
```

File: mlir/test/Target/SPIRV/function-decorations.mlir
Comment (kuhar) [inline mlir/test/Target/SPIRV/function-decorations.mlir:21]: Are the `// check` lines below even executed when not in upper case?

Comment (sott0n) [reply mlir/test/Target/SPIRV/function-decorations.mlir:21]: Sorry..these were in lower case by mistake.

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td
Comment (sott0n) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td]: I think it's not currently feasible to specify more than one decoration in the argument. Is there a method to provide both decorations within the argument?

Comment (kuhar) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td]: I thought the syntax was `spirv.decoration = #spirv.decoration<Aliased|Restricted>`, but if that's not accepted then it's fine as-is. Please double-check before merging.

**Review by kuhar — APPROVED** [review]
LGTM but please wait for a second approval from @antiagainst before merging

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td
Comment (sott0n) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVStructureOps.td]: I checked that this syntax　isn't accepted and causes an error below.

```
test.mlir:4:132: error: expected '>'
  spirv.func @access(%arg0 : !spirv.ptr<!spirv.struct<(f32)>, PhysicalStorageBuffer> { spirv.decoration = #spirv.decoration<Aliased|Restrict>, spirv.decoration = #spirv.decoration<Restrict>}) -> f32 "None" {
```

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.h
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Deserialization/Deserializer.h:617]: This should be a local variable in `processFunction` given this is only needed there and different functions have different argument attributes?

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:375]: Hmm, this assumes we always process `argID` in increasing order, which is the case for now. But not guaranteed. It's better to use `resize` if `argID >= argAttrs.size()`.

Also as it's implemented right now, we have `argAttrs` in the class directly. So this will just continuously grow if we process more than one functions. So it would be nice to add one more test in the serialization tests, where we have at least two functions, and each function has more than one arguments:

```
spirv.func @fn1(%arg0: i32, %arg1: !spirv.ptr<..> {spirv.decoration = ...})
spirv.func @fn2(%arg0: !spirv.ptr<..> {spirv.decoration = ...}, %arg1: !spirv.ptr<..> {spirv.decoration = ...})
```

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:388]: I'd think it's better to just break once found one of the above four decorations. We can create the dictionary attribute out of the for loop. This way we don't mutate `decorations` and it's clearer.

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.h
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Deserialization/Deserializer.h:237]: Rename to `setFunctionArgAttrs` to be clearer?

**Review by antiagainst — CHANGES_REQUESTED** [review]
Awesome, thanks for the contribution!

File: mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp
Comment (sott0n) [reply mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp:979]: Thanks for this comment! I remove it.

File: mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp
Comment (sott0n) [reply mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp:985]: Ok, I add checking the attribute name.

> It would be nice to add an IR test for such case to ignore #spirv.decoration attributes without spirv.decoration name.

I think that an attribute of an argument requires the decoration name, so this IR test expects an error `error: expected attribute name` for this case ( sets `#spirv.decoration` attributes without `spirv.decoration` name).

File: mlir/test/Dialect/SPIRV/IR/cast-ops.mlir
Comment (sott0n) [reply mlir/test/Dialect/SPIRV/IR/cast-ops.mlir:417]: > later we might want to have custom parser/printer to simpify it into something like %arg0 : !spirv.ptr<...> Aliased or something (cannot recall exactly the restrictions on attribute parsing/printing). 

It looks good! After applying this patch, I would like to think about the custom parse/printer.

File: mlir/lib/Target/SPIRV/Serialization/SerializeOps.cpp
Comment (sott0n) [reply mlir/lib/Target/SPIRV/Serialization/SerializeOps.cpp:191]: I checked this case in my local:

```bash
> cat test.mlir
spirv.module PhysicalStorageBuffer64 GLSL450 requires #spirv.vce<v1.0,
    [Shader, PhysicalStorageBufferAddresses], [SPV_KHR_physical_storage_buffer]> {
  spirv.func @func_arg_decoration_aliased(
      %arg0 : !spirv.ptr<i32, PhysicalStorageBuffer> { spirv.decoration = #spirv.decoration<Restrict>, random_attr = #spirv.decoration<Aliased> }
  ) "None" {
    spirv.Return
  }
}
```

This cases an error below:
```bash
> mlir-translate -no-implicit-module -test-spirv-roundtrip test.mlir
test.mlir:3:3: error: 'spirv.func' op arguments may only have dialect attributes
  spirv.func @func_arg_decoration_aliased(
  ^
test.mlir:3:3: note: see current operation:
"spirv.func"() <{arg_attrs = [{random_attr = #spirv.decoration<Aliased>, spirv.decoration = #spirv.decoration<Restrict>}], function_control = #spirv.function_control<None>, function_type = (!spirv.ptr<i32, PhysicalStorageBuffer>) -> (), sym_name = "func_arg_decoration_aliased"}> ({
^bb0(%arg0: !spirv.ptr<i32, PhysicalStorageBuffer>):
  "spirv.Return"() : () -> ()
}) : () -> ()

```

It seems to be not allowed the random attributes, but I agree with your comment. So I add checking a name of `spirv.decoration` before checking `DecorationAttr` and a test for random attribute case.

Comment (sott0n) [reply mlir/lib/Target/SPIRV/Serialization/SerializeOps.cpp:191]: Since this error is a verification on the IR side, I will add it to the test on the IR side, but is it necessary on the serialization/deserialization roundtrip side as well?

Comment (sott0n) [reply mlir/lib/Target/SPIRV/Serialization/SerializeOps.cpp:191]: Ok, I agree with you. So I add checking the name `spirv.decoration` before `dyn_cast<DecorationAttr>` and a test for this case.

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (sott0n) [reply mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:375]: > Also as it's implemented right now, we have argAttrs in the class directly. 

I move `argAttrs` into the `processFunction` as local variable.

> Hmm, this assumes we always process argID in increasing order, which is the case for now. But not guaranteed. It's better to use resize if argID >= argAttrs.size().

I understand that `argID` represents the ID for each variable's mapping, and it doesn't compare the length of `argAttrs`, correct? So my assumption is that the length of argAttrs matches the length of `getNumInputs`. In that case, why not reserve it as follows?

```cpp
LogicalResult
spirv::Deserializer::processFunction(ArrayRef<uint32_t> operands) {
  ...

  SmallVector<Attribute> argAttrs;
  argAttrs.reserve(functionType.getNumInputs());

  ...
```

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (sott0n) [reply mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:388]: Does this mean that during deserialization, only these four decorations should be added to `argAttrs` ?

File: mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp
Comment (antiagainst) [reply mlir/lib/Dialect/SPIRV/IR/SPIRVOps.cpp:985]: ah I meant adding a test like `%arg0: !spirv.ptr<..> { random_attr = #spirv.decoration<Aliased> }`.

File: mlir/test/Dialect/SPIRV/IR/cast-ops.mlir
Comment (antiagainst) [reply mlir/test/Dialect/SPIRV/IR/cast-ops.mlir:417]: SG!

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (antiagainst) [reply mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:375]: Yes `reserve` is helpful to allocate the space in one go. Though it does not initialize the elements--you'd still need to either `push_back` for all elements or call `resize` which initalizes elements along the way.

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (antiagainst) [reply mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:388]: Actually only one of these four for now. That reminds, we should error out if there are more we didn't recognize to avoid silently ignore decorations. So, make sure the decoration is only one of the four.

**Comment by antiagainst** [comment]
@sott0n I didn't see your fixes. Did you forget to push out or still working on it?

**Comment by sott0n** [comment]
@antiagainst Sorry.. I pushed it.

File: mlir/test/Target/SPIRV/function-decorations.mlir
Comment (antiagainst) [inline mlir/test/Target/SPIRV/function-decorations.mlir:104]: ah okay we have existing checks to error out for random attributes. that's good. :) we don't need tocheck error cases in the serialization test--these tests are meant valid cases only. (and the verification tests should be in the IR test directory, which you already have.) So fine to remove this and the above error case tests.

**Review by antiagainst — CHANGES_REQUESTED** [review]

File: mlir/test/Target/SPIRV/function-decorations.mlir
Comment (sott0n) [reply mlir/test/Target/SPIRV/function-decorations.mlir]: Thanks for your comment. I removed these tests.

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp
Comment (sott0n) [reply mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp]: Ok, I understand and updated this function.

> we should error out if there are more we didn't recognize to avoid silently ignore decorations.

Should we add this invalid test case?

Comment (antiagainst) [reply mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp]: I'm fine to do without a test for now, given that the possible error cases are unbound. Especially for the deserialization, right now it's meant to only cover the cases where we can serialize. 

BTW we do have some deserialization test here: https://github.com/llvm/llvm-project/blob/main/mlir/unittests/Dialect/SPIRV/DeserializationTest.cpp but it's a bit painful to write those.

**Review by antiagainst — APPROVED** [review]

**Comment by antiagainst** [comment]
Thanks a lot for the contribution and bearing with me for the nitpicking. :) To avoid burden you further, I rebased and revised some comment/error message slightly. :) I'll land once bots are happy.

**Comment by sott0n** [comment]
@kuhar @antiagainst Thanks for your comment and support to land this PR! If you have any issues or TODO tasks in SPIR-V dialect, I would be happy if you could share it with me.

**Comment by antiagainst** [comment]
> @kuhar @antiagainst Thanks for your comment and support to land this PR! If you have any issues or TODO tasks in SPIR-V dialect, I would be happy if you could share it with me. :)

Thanks for your further interest! Please feel free to take a look at issues with the `mlir:spirv` label to see if any one of them is interesting to you, especially those with the `good first issue` label at the same time!

**Pull request merged**
